File: q28-container-orchestrator/solution-1-java/ContainerManager.java
import java.util.*;

/**
 * In-memory Container Orchestrator.
 *
 * Features implemented per spec:
 * - Parse machine metadata: "machineId,totalCpuUnits,totalMemoryInMB".
 * - assignMachine(criteria, name, imageUrl, cpuUnits, memMb):
 *     * criteria==0 → pick machine with MAX spare CPU; tiebreak lexicographically by machineId.
 *     * criteria==1 → pick machine with MAX spare Memory; tiebreak lexicographically by machineId.
 *     * Must have free CPU ≥ cpuUnits and free Mem ≥ memMb.
 *     * Creates a RUNNING container on the chosen machine and reserves resources.
 *     * Returns machineId or "" if placement not possible.
 * - stop(name): RUNNING → STOPPED, frees CPU/Mem; true iff it existed, was assigned, and not already STOPPED.
 *
 * Notes:
 * - Containers start in RUNNING state on successful assignment.
 * - PAUSED state is defined for completeness (reserves resources like RUNNING),
 *   but no pause/resume APIs are required by this spec.
 * - All logic is single-threaded, in-memory, O(M) selection per placement.
 */
public class ContainerManager {

    private enum State { RUNNING, PAUSED, STOPPED }

    private static final class Machine {
        final String id;
        final int totalCpu;
        final int totalMem;
        int usedCpu;
        int usedMem;
        // Optional: track container names assigned here (not required by spec)
        final Set<String> containers = new HashSet<>();

        Machine(String id, int totalCpu, int totalMem) {
            this.id = id;
            this.totalCpu = totalCpu;
            this.totalMem = totalMem;
            this.usedCpu = 0;
            this.usedMem = 0;
        }

        int freeCpu() { return totalCpu - usedCpu; }
        int freeMem() { return totalMem - usedMem; }

        boolean hasCapacityFor(int cpu, int mem) {
            return freeCpu() >= cpu && freeMem() >= mem;
        }

        void allocate(int cpu, int mem, String containerName) {
            usedCpu += cpu;
            usedMem += mem;
            containers.add(containerName);
        }

        void release(int cpu, int mem, String containerName) {
            usedCpu -= cpu;
            usedMem -= mem;
            if (usedCpu < 0) usedCpu = 0;
            if (usedMem < 0) usedMem = 0;
            containers.remove(containerName);
        }
    }

    private static final class Container {
        final String name;
        final String imageUrl;
        final int cpu;
        final int mem;
        String machineId; // non-null only while RUNNING/PAUSED in this implementation
        State state;

        Container(String name, String imageUrl, int cpu, int mem, String machineId) {
            this.name = name;
            this.imageUrl = imageUrl;
            this.cpu = cpu;
            this.mem = mem;
            this.machineId = machineId;
            this.state = State.RUNNING;
        }
    }

    private final Map<String, Machine> machinesById = new HashMap<>();
    private final Map<String, Container> containersByName = new HashMap<>();

    /**
     * Constructor: parses machine rows of the form "machineId,totalCpuUnits,totalMemoryInMB".
     */
    public ContainerManager(List<String> machines) {
        if (machines == null) throw new IllegalArgumentException("machines list cannot be null");
        for (String row : machines) {
            if (row == null || row.isBlank()) {
                throw new IllegalArgumentException("Invalid machine row: " + row);
            }
            String[] parts = row.split(",", -1);
            if (parts.length != 3) {
                throw new IllegalArgumentException("Expected 3 fields: " + row);
            }
            String id = parts[0].trim();
            String cpuStr = parts[1].trim();
            String memStr = parts[2].trim();
            if (id.isEmpty()) throw new IllegalArgumentException("machineId cannot be blank: " + row);
            int cpu, mem;
            try {
                cpu = Integer.parseInt(cpuStr);
                mem = Integer.parseInt(memStr);
            } catch (NumberFormatException nfe) {
                throw new IllegalArgumentException("Non-integer capacity in row: " + row);
            }
            if (cpu <= 0 || mem <= 0) {
                throw new IllegalArgumentException("Capacities must be > 0: " + row);
            }
            if (machinesById.containsKey(id)) {
                throw new IllegalArgumentException("Duplicate machineId: " + id);
            }
            machinesById.put(id, new Machine(id, cpu, mem));
        }
    }

    /**
     * Assigns a machine to host a new container, using the selection criteria.
     * @param criteria 0 = max spare CPU; 1 = max spare Memory
     * @return machineId or "" if placement fails
     */
    public String assignMachine(int criteria, String containerName, String imageUrl, int cpuUnits, int memMb) {
        // Basic validation (per problem statement)
        if (containerName == null || containerName.isBlank()) return "";
        if (cpuUnits <= 0 || memMb <= 0) return "";
        if (containersByName.containsKey(containerName)) return ""; // enforce uniqueness defensively
        if (criteria != 0 && criteria != 1) return "";

        // Choose the best machine among those with sufficient capacity
        String chosenId = "";
        int bestMetric = Integer.MIN_VALUE;

        for (Machine m : machinesById.values()) {
            if (!m.hasCapacityFor(cpuUnits, memMb)) continue;

            int metric = (criteria == 0) ? m.freeCpu() : m.freeMem();
            if (metric > bestMetric) {
                bestMetric = metric;
                chosenId = m.id;
            } else if (metric == bestMetric && metric != Integer.MIN_VALUE) {
                // Tie-break: lexicographically smallest machineId
                if (!chosenId.isEmpty() && m.id.compareTo(chosenId) < 0) {
                    chosenId = m.id;
                } else if (chosenId.isEmpty()) {
                    chosenId = m.id;
                }
            }
        }

        if (chosenId.isEmpty()) {
            return "";
        }

        // Place the container
        Machine target = machinesById.get(chosenId);
        target.allocate(cpuUnits, memMb, containerName);
        Container c = new Container(containerName, imageUrl, cpuUnits, memMb, chosenId);
        containersByName.put(containerName, c);
        return chosenId;
    }

    /**
     * Stops a container: RUNNING/PAUSED → STOPPED, frees resources; returns true if state changed.
     */
    public boolean stop(String name) {
        if (name == null || name.isBlank()) return false;
        Container c = containersByName.get(name);
        if (c == null) return false;
        if (c.state == State.STOPPED) return false;
        if (c.machineId == null || c.machineId.isBlank()) return false; // not assigned (defensive)

        Machine host = machinesById.get(c.machineId);
        if (host == null) return false; // defensive, should not happen

        // Free resources and mark STOPPED
        host.release(c.cpu, c.mem, c.name);
        c.state = State.STOPPED;
        // Keep machineId for history; could be nulled if preferred:
        // c.machineId = null;
        return true;
    }

    // ---- Optional helpers for testing/inspection (not required by spec) ----

    /** Returns current free CPU and Mem for all machines, e.g. for debugging. */
    public Map<String, int[]> getMachineFreeResources() {
        Map<String, int[]> out = new TreeMap<>();
        for (Machine m : machinesById.values()) {
            out.put(m.id, new int[]{ m.freeCpu(), m.freeMem() });
        }
        return out;
    }

    /** Returns a container's state and host, if present. */
    public String getContainerInfo(String name) {
        Container c = containersByName.get(name);
        if (c == null) return "NOT_FOUND";
        return c.name + ":" + c.state + "@" + (c.machineId == null ? "-" : c.machineId);
    }
}

File: q28-container-orchestrator/solution-2-python/ContainerManager.py
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional


# In-memory Container Orchestrator (Python version of the provided Java reference).
# ------------------------------------------------------------------------------
# - Parse machine metadata rows: "machineId,totalCpuUnits,totalMemoryInMB".
# - assignMachine(criteria, name, imageUrl, cpuUnits, memMb):
#     * criteria == 0 → pick machine with MAX spare CPU; tiebreak lexicographically by machineId.
#     * criteria == 1 → pick machine with MAX spare Memory; tiebreak lexicographically by machineId.
#     * Must have free CPU ≥ cpuUnits and free Mem ≥ memMb.
#     * On success, create RUNNING container, reserve resources, return machineId; else return "".
# - stop(name): RUNNING/PAUSED → STOPPED, frees CPU/Mem; returns True iff it existed, assigned, and not already STOPPED.
# - Single-threaded, in-memory, O(M) selection per placement.
# ------------------------------------------------------------------------------


class State(Enum):
    RUNNING = "RUNNING"
    PAUSED = "PAUSED"
    STOPPED = "STOPPED"


@dataclass
class Machine:
    id: str
    total_cpu: int
    total_mem: int
    used_cpu: int = 0
    used_mem: int = 0
    containers: set = field(default_factory=set)  # container names hosted here

    # Compute free resources
    def free_cpu(self) -> int:
        return self.total_cpu - self.used_cpu

    def free_mem(self) -> int:
        return self.total_mem - self.used_mem

    # Capacity check
    def has_capacity_for(self, cpu: int, mem: int) -> bool:
        return self.free_cpu() >= cpu and self.free_mem() >= mem

    # Reserve resources for a container
    def allocate(self, cpu: int, mem: int, container_name: str) -> None:
        self.used_cpu += cpu
        self.used_mem += mem
        self.containers.add(container_name)

    # Release resources from a container
    def release(self, cpu: int, mem: int, container_name: str) -> None:
        self.used_cpu -= cpu
        self.used_mem -= mem
        if self.used_cpu < 0:  # defensive clamp
            self.used_cpu = 0
        if self.used_mem < 0:  # defensive clamp
            self.used_mem = 0
        self.containers.discard(container_name)


@dataclass
class Container:
    name: str
    image_url: str
    cpu: int
    mem: int
    machine_id: Optional[str]
    state: State = State.RUNNING  # starts RUNNING on successful placement


class ContainerManager:
    # Constructor: takes list of "machineId,totalCpuUnits,totalMemoryInMB"
    def __init__(self, machines: List[str]):
        self._machines_by_id: Dict[str, Machine] = {}
        self._containers_by_name: Dict[str, Container] = {}
        self.machines: List[str] = machines[:]  # keep original rows if needed by external tests

        if machines is None:
            raise ValueError("machines list cannot be None")

        for row in machines:
            if row is None or str(row).strip() == "":
                raise ValueError(f"Invalid machine row: {row}")
            parts = str(row).split(",", -1)
            if len(parts) != 3:
                raise ValueError(f"Expected 3 fields: {row}")
            mid = parts[0].strip()
            if mid == "":
                raise ValueError(f"machineId cannot be blank: {row}")
            try:
                cpu = int(parts[1].strip())
                mem = int(parts[2].strip())
            except Exception as e:
                # Use a generic ValueError per spec; test inputs are expected valid.
                raise ValueError(f"Non-integer capacity in row: {row}") from e
            if cpu <= 0 or mem <= 0:
                raise ValueError(f"Capacities must be > 0: {row}")
            if mid in self._machines_by_id:
                raise ValueError(f"Duplicate machineId: {mid}")

            self._machines_by_id[mid] = Machine(id=mid, total_cpu=cpu, total_mem=mem)

    # Assigns a machine for a new container using the given criteria.
    # Returns machineId or "" if placement fails.
    def assignMachine(self, criteria: int, containerName: str, imageUrl: str, cpuUnits: int, memMb: int) -> str:
        # Basic validations per problem statement
        if containerName is None or str(containerName).strip() == "":
            return ""
        if cpuUnits <= 0 or memMb <= 0:
            return ""
        if containerName in self._containers_by_name:  # enforce global uniqueness
            return ""
        if criteria not in (0, 1):
            return ""

        chosen_id = ""
        best_metric: Optional[int] = None

        # Iterate all machines and choose based on metric with tie-breaker
        for m in self._machines_by_id.values():
            if not m.has_capacity_for(cpuUnits, memMb):
                continue

            metric = m.free_cpu() if criteria == 0 else m.free_mem()

            if best_metric is None or metric > best_metric:
                best_metric = metric
                chosen_id = m.id
            elif metric == best_metric:
                # lexicographically smallest machineId
                if m.id < chosen_id:
                    chosen_id = m.id

        if chosen_id == "":
            return ""

        # Allocate on the chosen machine and register container
        host = self._machines_by_id[chosen_id]
        host.allocate(cpuUnits, memMb, containerName)
        cont = Container(
            name=containerName,
            image_url=imageUrl,
            cpu=cpuUnits,
            mem=memMb,
            machine_id=chosen_id,
            state=State.RUNNING,
        )
        self._containers_by_name[containerName] = cont
        return chosen_id

    # Stops a container: RUNNING/PAUSED → STOPPED; frees resources; returns True if state changed.
    def stop(self, name: str) -> bool:
        if name is None or str(name).strip() == "":
            return False
        c = self._containers_by_name.get(name)
        if c is None:
            return False
        if c.state == State.STOPPED:
            return False
        if not c.machine_id:
            return False  # not assigned (defensive)

        host = self._machines_by_id.get(c.machine_id)
        if host is None:
            return False  # defensive; should not happen

        host.release(c.cpu, c.mem, c.name)
        c.state = State.STOPPED
        # Keep machine_id for history; can be set to None if desired:
        # c.machine_id = None
        return True

    # -------- Optional helpers for testing/inspection (not part of the required API) --------

    # Returns sorted dict of machineId -> (freeCpu, freeMem)
    def getMachineFreeResources(self) -> Dict[str, List[int]]:
        out: Dict[str, List[int]] = {}
        for mid in sorted(self._machines_by_id.keys()):
            m = self._machines_by_id[mid]
            out[mid] = [m.free_cpu(), m.free_mem()]
        return out

    # Returns a string summary for a container
    def getContainerInfo(self, name: str) -> str:
        c = self._containers_by_name.get(name)
        if c is None:
            return "NOT_FOUND"
        return f"{c.name}:{c.state.value}@{c.machine_id if c.machine_id else '-'}"
